src/sshagent/OpenSSHKey.cpp

```diff
@@ -0,0 +1,481 @@
+{
+    m_privateData = data;
+}
+
+void OpenSSHKey::setComment(const QString& comment)
+{
+    m_comment = comment;
+}
+
+void OpenSSHKey::clearPrivate()
+{
+    m_rawPrivateData.clear();
+    m_privateData.clear();
+}
+
+bool OpenSSHKey::parsePEM(const QByteArray& in, QByteArray& out)
+{
+    QString pem = QString::fromLatin1(in);
+    QStringList rows = pem.split(QRegularExpression("[\r\n]"), QString::SkipEmptyParts);
+
+    if (rows.length() < 3) {
+        m_error = tr("Invalid key file, expecting an OpenSSH key");
+        return false;
+    }
+
+    QString begin = rows.first();
+    QString end = rows.last();
+
+    QRegularExpressionMatch beginMatch = QRegularExpression("^-----BEGIN ([^\\-]+)-----$").match(begin);
+    QRegularExpressionMatch endMatch = QRegularExpression("^-----END ([^\\-]+)-----$").match(end);
+
+    if (!beginMatch.hasMatch() || !endMatch.hasMatch()) {
+        m_error = tr("Invalid key file, expecting an OpenSSH key");
+        return false;
+    }
+
+    if (beginMatch.captured(1) != endMatch.captured(1)) {
+        m_error = tr("PEM boundary mismatch");
+        return false;
+    }
+
+    if (beginMatch.captured(1) != "OPENSSH PRIVATE KEY") {
+        m_error = tr("This is not an OpenSSH key, only modern keys are supported");
+        return false;
+    }
+
+    rows.removeFirst();
+    rows.removeLast();
+
+    out = QByteArray::fromBase64(rows.join("").toLatin1());
+
+    if (out.isEmpty()) {
+        m_error = tr("Base64 decoding failed");
+        return false;
+    }
+
+    return true;
+}
+
+bool OpenSSHKey::parse(const QByteArray& in)
+{
+    QByteArray data;
+
+    if (!parsePEM(in, data)) {
+        return false;
+    }
+
+    BinaryStream stream(&data);
+
+    QByteArray magic;
+    magic.resize(15);
+
+    if (!stream.read(magic)) {
+        m_error = tr("Key file way too small.");
+        return false;
+    }
+
+    if (QString::fromLatin1(magic) != "openssh-key-v1") {
+        m_error = tr("Key file magic header id invalid");
+        return false;
+    }
+
+    stream.readString(m_cipherName);
+    stream.readString(m_kdfName);
+    stream.readString(m_kdfOptions);
+
+    quint32 numberOfKeys;
+    stream.read(numberOfKeys);
+
+    if (numberOfKeys == 0) {
+        m_error = tr("Found zero keys");
+        return false;
+    }
+
+    for (quint32 i = 0; i < numberOfKeys; i++) {
+        QByteArray publicKey;
+        if (!stream.readString(publicKey)) {
+            m_error = tr("Failed to read public key.");
+            return false;
+        }
+
+        if (i == 0) {
+            BinaryStream publicStream(&publicKey);
+            if (!readPublic(publicStream)) {
+                return false;
+            }
+        }
+    }
+
+    // padded list of keys
+    if (!stream.readString(m_rawPrivateData)) {
+        m_error = tr("Corrupted key file, reading private key failed");
+        return false;
+    }
+
+    // load private if no encryption
+    if (!encrypted()) {
+        return decrypt();
```

Comment:
Am I missing something? Why do you decrypt if it is not encrypted?

Reply:
Decryption also implies parsing the private portion regardless if there's actual encryption used. Function naming issue.

Reply:
OK. Should be renamed then.